### PR TITLE
[ENG-3945]Validate DOIs on PATCH

### PR DESCRIPTION
**Pull request by jwalz**
## Purpose
We want to alert users when they try to assign an invalid DOI to a resource. For DOIs, we do this by curling against `https://doi.or/ra/<doi_value>`. This will tell is where the DOI is registered if it is valid. Otherwise, it will tell us if the DOI is either malformed or simply hasn't been registered.  We skip this step for DOIs that we generated (i.e. DOIs that have an OSF `referent`), since we have a high level of confidence in their correctness.

Pinging the `/ra/` endpoint also empowers future work to store the registering agency to enable querying of PID metadata in the future.

## Changes
* Add a `PIDValidator` Abstract class and `DOIValidator` subclass to handle validation for `Identifier` values
  * Add `settings` values to control if validation happens and what endpoints to use
  * Add informative exceptions for different types of errors
* Add `Identifier.validate_identifier_value` and call it when `OutcomeArtifact.update_identifier` creates a new identifier
* Update error handling in `ResourceSerializer.update` to handle the new exceptions
* Tests

The `PIDValidator` functionality is prematurely abstractified in order to lay explicit groundwork for expansion to future PID types.

## QA Notes
The DOIs we generate for Registrations on test are only registered with Datacite's test server (i.e. doi.org does not know about them) and as such will fail validation. We have to use real-live DOIs to try this out where `settings.DOI_VALIDATION_ENABLED` is True.

## Ticket
https://openscience.atlassian.net/browse/ENG-3945

### CODE REVIEW

File: osf/exceptions.py [hunk 1]
```diff
@@ -229,10 +229,32 @@ class IdentifierHasReferencesError(OSFError):
...
+    def __init__(self, message):
+        self.message = message
+
+
+class InvalidPIDFormatError(InvalidPIDError):
+    ERROR_MESSAGE = '{value} does not follow the proper formatting for a PID of type {category}'
+
+
+class NoSuchPIDError(InvalidPIDError):
+    ERROR_MESSAGE = 'Could not find any record of PID of type {category} with valuea {value}'
```
Comment (brianjgeiger) [inline osf/exceptions.py:255]:
```suggestion
    ERROR_MESSAGE = 'Could not find any record of PID of type {category} with value {value}'
```

File: osf/utils/identifiers.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+        # These error messages were copied from actual responses;
+        # If they change, still raise an error, just not the most descriptive one
+        error_status = response_data.get('status')
+        if error_status == 'DOI does not exist':
+            pid_exception = NoSuchPIDError
+        elif error_status == 'Invalid DOI':
+            pid_exception = InvalidPIDFormatError
+        else:
+            pid_exception = InvalidPIDError
```
Comment (brianjgeiger) [inline osf/utils/identifiers.py:72]: Would it be worth logging a sentry error in this case so we know, hypothetically, if they change their error messages, and we could update the code to catch the new error?

**Review by brianjgeiger — CHANGES_REQUESTED** [review]
One typo, one suggestion, and one question.

File: osf/utils/identifiers.py (hunk 2 above)
Comment (jwalz) [reply osf/utils/identifiers.py:72]: Yes, yes it would.

File: website/settings/defaults.py [hunk 3]
```diff
@@ -320,6 +320,11 @@ def parent_dir(path):
 ####################
 #   Identifiers   #
 ###################
+PID_VALIDATION_ENABLED = False
+PID_VALIDATION_ENDPOINTS = {
```
[the remark this answers is not in the document]
Comment (jwalz) [reply website/settings/defaults.py:324]: A bunch of things that made sense in the moment led to this decision 😅 

1) I was originally not going to have the top-level PID_VALIDATION_ENABLED setting and just set this to '' to turn it off per-environment, but then I realized that was a terribly confusing idea

2) I really hoped that we would be able to validate our test data in a way that was congruent with how we're validating our real data and could just change the value here

3) I still hope that (2) will be true for other PID types

That said, this is probably the most premature generalization here, and I'm happy to just make it a member on the validator class.

[the remark this answers is not in the document]
Comment (brianjgeiger) [reply website/settings/defaults.py:324]: Eh. It doesn't hurt anything and does give us some options. Unless Matt hates it, I think it's fine.

**Review by brianjgeiger — APPROVED** [review]

File: osf/models/identifiers.py [hunk 4]
```diff
@@ -36,6 +40,20 @@ def delete(self):
...
+    def validate_identifier_value(self):
+        # We created the PID, so assume valid
+        if self.object_id is not None:
+            return True
+
+        # If we don't know how to validate a PID, assume it's fine
+        try:
+            validator = identifier_utils.PIDValidator.for_identifier_category(self.category)
+        except NoSuchPIDValidatorError:
+            return True
```
Comment (mfraezz) [inline osf/models/identifiers.py:52]: Minor: Should this be logged in any way, e.g. `sentry.log_message(f'Unvalidatable PID category detected: {self.category}')`

File: website/settings/defaults.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mfraezz) [reply website/settings/defaults.py:324]: I think it's fine.

**Review by mfraezz — APPROVED** [review]
LGTM. One minor question, but no blockers. :octocat:

File: osf/models/identifiers.py (hunk 4 above)
Comment (jwalz) [reply osf/models/identifiers.py:52]: Added sentry logging in `PIDValidator` prior to raising

**Pull request merged**